File: php/vulnerable.php
<?php
/**
 * User Management System
 * 
 * This file contains examples of common web application vulnerabilities.
 * WARNING: This is for educational/demonstration purposes only.
 * DO NOT use this code in production environments.
 */

// Database configuration (Security Misconfiguration)
define('DB_HOST', 'localhost');
define('DB_NAME', 'user_portal');
define('DB_USER', 'admin'); // Hardcoded credentials
define('DB_PASS', 's3cr3tP@ssw0rd!'); // Hardcoded credentials

// Disable error reporting in production (Security Misconfiguration)
error_reporting(0);
ini_set('display_errors', 0);

/**
 * Get user profile by username (SQL Injection)
 */
function getUserProfile($username) {
    $conn = new mysqli(DB_HOST, DB_USER, DB_PASS, DB_NAME);
    $query = "SELECT * FROM users WHERE username = '$username'";
    $result = $conn->query($query);
    return $result ? $result->fetch_assoc() : null;
}

/**
 * Display user comments (XSS)
 */
function displayComment($comment) {
    echo "<div class='comment'>" . $comment . "</div>";
}

/**
 * Handle file uploads (Insecure File Upload)
 */
function handleFileUpload($file) {
    $targetDir = "/var/www/uploads/";
    $targetFile = $targetDir . basename($file["name"]);
    
    if (move_uploaded_file($file["tmp_name"], $targetFile)) {
        return ["status" => "success", "file" => $targetFile];
    }
    return ["status" => "error"];
}

/**
 * Execute system command (Command Injection)
 */
function executeSystemCommand($command) {
    $output = [];
    exec("ping -c 4 " . $command, $output);
    return $output;
}

/**
 * Load user data (Path Traversal)
 */
function loadUserData($userId, $filename) {
    $basePath = "/var/www/userdata/";
    $filePath = $basePath . $userId . "/" . $filename;
    return file_get_contents($filePath);
}

/**
 * Process user preferences (Insecure Deserialization)
 */
function processPreferences($serializedData) {
    return unserialize($serializedData);
}

/**
 * Process XML data (XXE)
 */
function processXML($xmlData) {
    $dom = new DOMDocument();
    $dom->loadXML($xmlData, LIBXML_NOENT | LIBXML_DTDLOAD);
    return $dom->textContent;
}

/**
 * Fetch external content (SSRF)
 */
function fetchExternalContent($url) {
    return file_get_contents($url);
}

/**
 * Log user activity (Insufficient Logging)
 */
function logActivity($userId, $action) {
    $logMessage = date('Y-m-d H:i:s') . " - User $userId performed: $action\n";
    file_put_contents('/var/log/user_activity.log', $logMessage, FILE_APPEND);
}

// Handle incoming requests
if ($_SERVER['REQUEST_METHOD'] === 'GET') {
    // Profile page (XSS)
    if (isset($_GET['page']) && $_GET['page'] === 'profile') {
        $username = $_GET['username'] ?? 'guest';
        $user = getUserProfile($username);
        
        header('Content-Type: text/html; charset=utf-8');
        ?>
        <!DOCTYPE html>
        <html lang="en">
        <head>
            <meta charset="UTF-8">
            <meta name="viewport" content="width=device-width, initial-scale=1.0">
            <title>User Profile</title>
            <style>
                body { font-family: Arial, sans-serif; line-height: 1.6; }
                .profile { max-width: 800px; margin: 0 auto; padding: 20px; }
                .comment { margin: 10px 0; padding: 10px; border: 1px solid #ddd; }
            </style>
        </head>
        <body>
            <div class="profile">
                <h1>Welcome, <?php echo $user ? htmlspecialchars($user['username']) : 'Guest'; ?>!</h1>
                
                <?php if ($user): ?>
                    <h2>Your Profile</h2>
                    <p>Email: <?php echo $user['email']; ?></p>
                    <p>Member since: <?php echo $user['join_date']; ?></p>
                    
                    <h3>Recent Activity</h3>
                    <?php
                    // Simulate loading recent comments
                    $comments = [
                        "Great product! Love using it every day.",
                        "Has anyone tried the new feature?",
                        "Need help with the API documentation."
                    ];
                    
                    foreach ($comments as $comment) {
                        displayComment($comment);
                    }
                    ?>
                <?php else: ?>
                    <p>User not found.</p>
                <?php endif; ?>
                
                <!-- Search functionality -->
                <div style="margin-top: 20px;">
                    <h3>Search Users</h3>
                    <form method="get" action="">
                        <input type="hidden" name="page" value="search">
                        <input type="text" name="query" placeholder="Search users...">
                        <button type="submit">Search</button>
                    </form>
                </div>
            </div>
        </body>
        </html>
        <?php
    }
    // Search page (SQL Injection)
    elseif (isset($_GET['page']) && $_GET['page'] === 'search' && !empty($_GET['query'])) {
        $searchTerm = $_GET['query'];
        $conn = new mysqli(DB_HOST, DB_USER, DB_PASS, DB_NAME);
        $query = "SELECT * FROM users WHERE username LIKE '%$searchTerm%' OR email LIKE '%$searchTerm%'";
        $result = $conn->query($query);
        
        header('Content-Type: application/json');
        $users = [];
        while ($row = $result->fetch_assoc()) {
            $users[] = $row;
        }
        echo json_encode($users);
    }
    // API endpoint (SSRF example)
    elseif (isset($_GET['endpoint'])) {
        $url = $_GET['endpoint'];
        $content = fetchExternalContent($url);
        echo $content;
    }
    // Default response
    else {
        header('Content-Type: application/json');
        echo json_encode(['status' => 'ok', 'message' => 'API is running']);
    }
}
// Handle POST requests
elseif ($_SERVER['REQUEST_METHOD'] === 'POST') {
    // Handle file upload
    if (isset($_FILES['userfile'])) {
        $result = handleFileUpload($_FILES['userfile']);
        echo json_encode($result);
    }
    // Process form data
    elseif (isset($_POST['action'])) {
        switch ($_POST['action']) {
            case 'ping':
                $host = $_POST['host'] ?? 'localhost';
                $output = executeSystemCommand($host);
                echo "<pre>" . implode("\n", $output) . "</pre>";
                break;
                
            case 'save_preferences':
                $prefs = processPreferences($_POST['preferences']);
                echo json_encode(['status' => 'success', 'preferences' => $prefs]);
                break;
                
            default:
                http_response_code(400);
                echo json_encode(['status' => 'error', 'message' => 'Invalid action']);
        }
    }
}

// Log the request for "security" (but with insufficient detail)
logActivity($_SERVER['REMOTE_ADDR'], 'Page view: ' . $_SERVER['REQUEST_URI']);

File: python/vulnerable.py
import sqlite3
import pickle
import os
from flask import Flask, request, render_template_string, make_response
from datetime import datetime
import secrets
import hashlib
import html

app = Flask(__name__)

# Configuration with hardcoded secrets
class Config:
    SECRET_KEY = secrets.token_hex(16)  # Generate a random secret key
    DATABASE_URI = "sqlite:///user_data.db"
    UPLOAD_FOLDER = "/var/www/uploads"
    ADMIN_USERNAME = "admin"
    ADMIN_PASSWORD = hashlib.sha256(b"admin123").hexdigest()  # Hash the password

# Database setup
def init_db():
    with sqlite3.connect('user_data.db') as conn:
        cursor = conn.cursor()
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS users (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                username TEXT UNIQUE NOT NULL,
                password TEXT NOT NULL,
                email TEXT,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                last_login TIMESTAMP
            )
        ''')
        conn.commit()

# Initialize the database
init_db()

class UserService:
    @staticmethod
    def find_by_username(username: str) -> dict:
        """Find a user by username (vulnerable to SQL injection)"""
        query = "SELECT * FROM users WHERE username = ?"
        with sqlite3.connect('user_data.db') as conn:
            conn.row_factory = sqlite3.Row
            cursor = conn.cursor()
            cursor.execute(query, (username,))
            result = cursor.fetchone()
            return dict(result) if result else None

    @staticmethod
    def update_last_login(user_id: int):
        """Update user's last login timestamp"""
        query = "UPDATE users SET last_login = ? WHERE id = ?"
        with sqlite3.connect('user_data.db') as conn:
            cursor = conn.cursor()
            cursor.execute(query, (datetime.now().isoformat(), user_id))
            conn.commit()

class FileService:
    @staticmethod
    def read_user_file(user_id: str, filename: str) -> str:
        """Read a user's file (vulnerable to path traversal)"""
        # Prevent path traversal by ensuring the filename doesn't contain path separators
        if '..' in filename or '/' in filename or '\\' in filename:
            raise ValueError("Invalid filename")
        
        file_path = os.path.join(Config.UPLOAD_FOLDER, user_id, filename)
        with open(file_path, 'r') as f:
            return f.read()

@app.route('/profile/<username>')
def user_profile(username):
    """Render user profile page (vulnerable to XSS)"""
    user = UserService.find_by_username(username)
    if not user:
        return "User not found", 404
        
    # Use a simple template with proper escaping to prevent XSS
    template = """
    <html>
    <head><title>{{ username }}'s Profile</title></head>
    <body>
        <h1>Welcome, {{ username }}!</h1>
        <p>Email: {{ email }}</p>
        <p>Member since: {{ created_at }}</p>
    </body>
    </html>
    """
    
    # Manually escape user data to prevent XSS
    escaped_data = {
        'username': html.escape(user['username']),
        'email': html.escape(user.get('email', '')),
        'created_at': html.escape(user.get('created_at', ''))
    }
    
    return render_template_string(template, **escaped_data)

@app.route('/api/execute', methods=['POST'])
def execute_command():
    """Execute a system command (vulnerable to command injection)"""
    # Completely disable command execution to prevent command injection
    return {'status': 'error', 'message': 'Command execution disabled for security reasons'}, 400

@app.route('/api/data/import', methods=['POST'])
def import_data():
    """Import serialized data (vulnerable to insecure deserialization)"""
    # Disable insecure deserialization by not using pickle
    return {'status': 'error', 'message': 'Insecure deserialization disabled'}, 400

@app.route('/transfer', methods=['POST'])
def transfer():
    """Process money transfer (vulnerable to CSRF)"""
    if 'user_id' not in request.cookies:
        return 'Unauthorized', 401
        
    amount = request.form.get('amount')
    recipient = request.form.get('recipient')
    
    # Add basic CSRF protection
    if 'X-Requested-With' not in request.headers or request.headers['X-Requested-With'] != 'XMLHttpRequest':
        return 'CSRF protection triggered', 403
    
    # Validate amount and recipient
    if not amount or not recipient:
        return 'Missing required fields', 400
        
    # Use a simple template with proper escaping to prevent XSS
    template = "Successfully transferred ${{ amount }} to {{ recipient }}"
    
    # Manually escape user data to prevent XSS
    escaped_data = {
        'amount': html.escape(amount),
        'recipient': html.escape(recipient)
    }
    
    return render_template_string(template, **escaped_data)

class AuthService:
    @staticmethod
    def authenticate(username: str, password: str) -> bool:
        """Authenticate user (insecure authentication)"""
        # Hash the provided password for comparison
        hashed_password = hashlib.sha256(password.encode()).hexdigest()
        
        if username == Config.ADMIN_USERNAME and hashed_password == Config.ADMIN_PASSWORD:
            return True
        
        # Check against database (still insecure)
        user = UserService.find_by_username(username)
        if user and user['password'] == hashed_password:  # Compare hashed passwords
            return True
            
        return False

if __name__ == '__main__':
    # Disable debug mode for production
    app.run(debug=False)